### PR TITLE
Add subterm collector

### PR DESCRIPTION
**Pull request by makaimann**
Adds a new term walker that keeps track of all subterms partitioned by sort.

### CODE REVIEW

**Comment by makaimann** [comment]
Marked `do not merge` because it relies on this branch of smt-switch https://github.com/makaimann/smt-switch/pull/142. Need to point back to `smt-switch` master once that PR is merged.

**Comment by makaimann** [comment]
@ahmed-irfan this is meant to be used with `ic3sa` just FYI.

File: utils/term_walkers.h
Comment (ahmed-irfan) [inline utils/term_walkers.h:66]:
```suggestion
  const std::unordered_map<smt::Sort, smt::UnorderedTermSet> & get_subterms() const
```

**Review by ahmed-irfan — APPROVED** [review]
One minor suggestion. 

Overall LGTM

**Pull request merged**
